src/Catalyst.Core.Lib/Dag/DagNode.cs

```diff
@@ -0,0 +1,379 @@
+                stream.WriteLength(DataBytes.Length);
+                stream.WriteSomeBytes(DataBytes);
+            }
+        }
+
+        private void Read(Stream stream)
+        {
+            using (var cis = new CodedInputStream(stream, true))
+            {
+                Read(cis);
+            }
+        }
+
+        private void Read(CodedInputStream stream)
+        {
+            var links = new List<DagLink>();
+            var done = false;
+
+            while (!stream.IsAtEnd && !done)
+            {
+                var tag = stream.ReadTag();
+                switch (WireFormat.GetTagFieldNumber(tag))
+                {
+                    case 1:
+                        DataBytes = stream.ReadSomeBytes(stream.ReadLength());
+                        done = true;
+                        break;
+                    case 2:
+                        using (var ms = new MemoryStream(stream.ReadSomeBytes(stream.ReadLength())))
+                        {
+                            links.Add(new DagLink(ms));
+                        }
+
+                        break;
+                    default:
+                        throw new InvalidDataException("Unknown field number");
+                }
+            }
+
+            if (DataBytes == null)
+                DataBytes = new byte[0];
+            Links = links.ToArray();
+        }
+
+        /// <summary>
+        ///     Returns the IPFS binary representation as a byte array.
+        /// </summary>
+        /// <returns>
+        ///     A byte array.
+        /// </returns>
+        public byte[] ToArray()
+        {
+            using (var ms = new MemoryStream())
+            {
+                Write(ms);
+                return ms.ToArray();
+            }
+        }
+
+        private void ComputeHash()
+        {
+            using (var ms = new MemoryStream())
+            {
+                Write(ms);
+                size = ms.Position;
+                ms.Position = 0;
+                id = MultiHash.ComputeHash(ms, hashAlgorithm);
+            }
+        }
+
+        private void ComputeSize()
+        {
+            using (var ms = new MemoryStream())
+            {
+                Write(ms);
+                size = ms.Position;
+            }
+        }
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+        {
+            var that = obj as DagNode;
+            return that != null && Id == that.Id;
+        }
+
+        /// <inheritdoc />
+        public bool Equals(DagNode that) { return that != null && Id == that.Id; }
+
+        /// <summary>
+        ///     TODO
+        /// </summary>
+        public static bool operator ==(DagNode a, DagNode b)
+        {
+            if (ReferenceEquals(a, b)) return true;
+            if (ReferenceEquals(a, null)) return false;
+            if (ReferenceEquals(b, null)) return false;
+
+            return a.Equals(b);
+        }
+
+        /// <summary>
+        ///     TODO
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)